[Cases] Filter user activities pagination

js-jankisalvi opened the pull request:
## Summary

Implements #130227


https://user-images.githubusercontent.com/117571355/224038340-6b1c8cc7-3795-4412-8e67-a026dfe10776.mov



### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

### Flaky Test Runner
https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/2059

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

elasticmachine commented:
Pinging @elastic/response-ops-cases (Feature:Cases)

cnasikas commented on x-pack/plugins/cases/public/components/case_view/components/case_view_activity.tsx:
```diff
@@ -297,7 +290,7 @@ export const CaseViewActivity = ({
               theCase={caseData}
               headline={i18n.PARTICIPANTS}
-              loading={isLoadingUserActions}
+              loading={isLoadingCaseUsers}
```
Probably a mistake 🙂

cnasikas reviewed (commented):
Amazing job! It is so exciting seeing pagination in the user actions. I did a quick pass and I have some comments:

- I think the `UserActionsList` should not be aware of the "Show more" button. The parent component (`UserActions`) should render the "Show more" button and handle all the logic.
- The fetching should be moved from the `UserActionsList` to `UserActions`. The `UserActionsList` should have only one job, to build and render the user actions (passed as props)
- We should create a hook (`useUserActionsPagination` or similar) to encapsulate the logic about the pagination (page, last page calculation, etc). The hook can also do the calls to `useInfiniteFindCaseUserActions` and `useFindCaseUserActions` and export what is needed (`handleShowMore` etc.). The parent can use the hook to render the lists.

js-jankisalvi commented:
> * I think the `UserActionsList` should not be aware of the "Show more" button. The parent component (`UserActions`) should render the "Show more" button and handle all the logic.
Updated. Also added css suggested by Eui team:
![image](https://user-images.githubusercontent.com/117571355/224356666-b22afb32-6f13-408d-98ab-77a600f1f06c.png)

> * We should create a hook (`useUserActionsPagination` or similar) to encapsulate the logic about the pagination (page, last page calculation, etc). The hook can also do the calls to `useInfiniteFindCaseUserActions` and `useFindCaseUserActions` and export what is needed (`handleShowMore`, `data`, `showBottonList` etc.). The parent can use the hook to render the lists.
Moved to `use_user_actions_pagination` hook.

jonathan-buttner reviewed (commented):
Looking good! I left a few questions.

Couple questions:

When I refresh the case I see the `_find` getting called 4 times for some reason 🤔 

![image](https://user-images.githubusercontent.com/56361221/224414345-c03b2f30-7b08-445e-8912-4872ca502a99.png)

When I click the load more button it seems to request the first and last pages. Those should be cached right?

It also seems to make requests for each page, for example I click load more a couple of times and the last time I clicked it, dev tools shows requests for all previous pages

![image](https://user-images.githubusercontent.com/56361221/224416836-87bbc3f6-8158-4be7-92bc-012cfb62d957.png)

cnasikas commented on x-pack/plugins/cases/public/components/user_actions/index.tsx:
```diff
@@ -35,24 +38,27 @@ export const UserActions = React.memo((props: UserActionTreeProps) => {
-    }
-  }, [userActionsStats, userActivityQueryParams]);
+  const { caseUserActions, isLoadingUserActions } = useUserActionsPagination({
+    userActivityQueryParams: { ...userActivityQueryParams, page: lastPage },
```
I don't think we need two calls of the `useUserActionsPagination`. We can combine the needed logic into one call. Also by doing two calls of the `useUserActionsPagination` call we also do four calls to the backend two for each invocation.

js-jankisalvi replied:
Good point!

cnasikas commented on x-pack/plugins/cases/public/components/user_actions/index.tsx:
```diff
@@ -99,31 +110,106 @@ export const UserActions = React.memo((props: UserActionTreeProps) => {
+          isExpandable
+        />
+        {showLoadMore && hasNextPage && (
+          <EuiPanel
+            color="subdued"
+            css={css`
+              display: flex;
+              justify-content: center;
+              margin-block: ${euiTheme.size.base};
+              margin-inline-start: ${euiTheme.size.xxxl};
+            `}
+          >
+            <EuiButton
+              fill
+              color="text"
+              size="s"
+              onClick={handleShowMore}
+              data-test-subj="show-more-user-actions"
+            >
+              {i18n.SHOW_MORE}
+            </EuiButton>
+          </EuiPanel>
```
Let's create a component for the button. For example `ShowMoreButton`.

js-jankisalvi commented:
> When I refresh the case I see the `_find` getting called 4 times for some reason 🤔
Hmm.. should be 2 calls only, will take a look.
> When I click the load more button it seems to request the first and last pages. Those should be cached right?
yes, they should be cached.
> It also seems to make requests for each page, for example I click load more a couple of times and the last time I clicked it, dev tools shows requests for all previous pages
Looks like the `pagination hook `is called more than I intend to 🤦

js-jankisalvi commented on x-pack/plugins/cases/public/components/case_view/components/case_view_activity.test.tsx:
```diff
@@ -139,23 +148,48 @@ describe.skip('Case View Page activity tab', () => {
+
+    const lastPageForAll = Math.ceil(userActionsStats.total / userActivityQueryParams.perPage);
+
+    expect(useFindCaseUserActionsMock).toHaveBeenCalledWith(
```
hmm.. good point, I will put a check for number of times it is called.

js-jankisalvi replied:
Removed isEnabled option from parent and handled it inside `pagination` hook to enable or disable calls based on input props change.

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/use_user_actions_pagination.tsx:
```diff
@@ -0,0 +1,103 @@
+  caseId,
+  isExpandable = false,
+}: UserActionsPagination) => {
+  const isFirstRender = useRef(true);
```
It's a leftover variable, not needed anymore.

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/use_user_actions_pagination.tsx:
```diff
@@ -0,0 +1,103 @@
+    fetchNextPage,
+  } = useInfiniteFindCaseUserActions(caseId, userActivityQueryParams, isExpandable);
+
+  const { data: caseUserActionsData, isLoading: isLoadingUserActions } = useFindCaseUserActions(
```
`useInfiniteFindCaseUserActions ` uses useInfiniteQuery hook which is for pagination. 
`useFindCaseUserActions` uses normal useQuery hook, which helps to find last page result as the first hook will not return last page directly. I am currently omitting last page call in the ` getNextPageParam` option  from `useInfiniteFindCaseUserActions`/
I will have to check if I can put some logic to fetch last page directly from  `useInfiniteFindCaseUserActions`

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/index.tsx:
```diff
@@ -5,251 +5,201 @@
+    caseId: caseData.id,
+  });
+
+  const { isLoadingLastPageUserActions, lastPageUserActions } = useLastPageUserActions({
```
created separate hook for last page user actions for clear separation of concerns as it was confusing for me to have it in pagination hook.

jonathan-buttner reviewed (commented):
Looking good! Left a couple comments. I don't see the multiple calls to the find api anymore though 👍

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/use_user_actions_last_page.test.tsx:
```diff
@@ -0,0 +1,324 @@
+        userActivityQueryParams: { ...userActivityQueryParams, sortOrder: 'desc' },
+      });
+
+      expect(useFindCaseUserActionsMock).toHaveBeenCalledWith(
```
I am checking that when this rerender happens, it will set `enabled` of fetch API hook to false at line 320.

js-jankisalvi replied:
removed useRefs and rerender tests.

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/index.tsx:
```diff
@@ -5,251 +5,203 @@
+          isExpandable
+        />
+        {showLoadMore && hasNextPage && <ShowMoreButton onShowMoreClick={handleShowMore} />}
+        {showBottomList ? (
```
yeah, ideally it should not complicate things to add this condition instead of `showBottomList`.

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/use_user_actions_last_page.tsx:
```diff
@@ -0,0 +1,84 @@
+  lastPage,
+  showBottomList,
+}: LastPageUserActions) => {
+  const isFirstRender = useRef(true);
```
This hook is called every time there is a change on parent component. So o avoid unnecessary calls to fetch API hook, I disabled rerender after First render of hook until any props is changed.

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/use_user_actions_last_page.tsx:
```diff
@@ -0,0 +1,84 @@
+  const isActionsStatsUpdated =
+    userActivityQueryParams.sortOrder === 'asc' &&
+    !deepEqual(actionsStats.current, userActionsStats) &&
+    !isFirstRender.current; // refetch last page actions only when new action added in the bottom list (i.e. in descending order)
```
My bad, yes I meant `ascending`.  
In Ascending, new actions will be added in the last page.

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/index.tsx:
```diff
@@ -162,44 +163,34 @@ export const UserActions = React.memo((props: UserActionTreeProps) => {
           commentRefs={commentRefs}
           handleManageQuote={handleManageQuote}
-          bottomActions={lastPage === 0 ? bottomActions : []}
+          bottomActions={lastPage === 1 ? bottomActions : []}
```
Because previously we had no actions at all, but now we have by default case created action for newly created case and due to which the lastPage is 1.
But I should add check for `0 `as well because for new case there are no comments, and lastPage for comments will be 0

js-jankisalvi commented on x-pack/test/functional_with_es_ssl/apps/cases/group1/view_case.ts:
```diff
@@ -377,52 +396,125 @@ export default ({ getPageObject, getService }: FtrProviderContext) => {
+      });
+
+      it('shows more actions on button click', async () => {
+        await cases.common.selectSeverity(CaseSeverity.LOW);
```
yes, I am adding different user actions so that User actions reach to more than 20 and then Show more button is visible. No it will automatically `reduce comments and show load more button` when 21st comment is added.

I will create separate functionality to generate user actions.

js-jankisalvi commented on x-pack/test/functional_with_es_ssl/apps/cases/group1/view_case.ts:
```diff
@@ -377,52 +396,125 @@ export default ({ getPageObject, getService }: FtrProviderContext) => {
         expect(await historyBadge.getVisibleText()).equal('3');
       });
 
-      it('filters by comment successfully', async () => {
```
It is there, just moved at the top at line 362. Did some rearranging for readability.

js-jankisalvi commented on x-pack/test/functional_with_es_ssl/apps/cases/group1/view_case.ts:
```diff
@@ -377,52 +396,125 @@ export default ({ getPageObject, getService }: FtrProviderContext) => {
-      });
+        await header.waitUntilLoadingHasFinished();
 
-      it('sorts by newest first successfully', async () => {
```
yup, it's a separate test at line 399.

cnasikas commented on x-pack/test/functional/services/cases/common.ts:
I think is best if you do an API call to generate the user actions. You can update the case (check `updateCase` in `x-pack/test/functional/services/cases/api.ts`) with the following body:

```
{
    "cases": [
        {
            "title": "my title",
            "description": "my desc",
            "severity": "medium",
            "status": "in-progress",
            // other fields if you want
            "id": <case_id>,
            "version": <case_version>
        }
    ]
}
```
If you want to generate more you can do multiple calls to the API. You can put it here `x-pack/test/functional/services/cases/api.ts`.

cnasikas replied:
Yes, each time you update the case the version changes. You need to get the new version from the latest response and put it in the next request.

cnasikas commented on x-pack/plugins/cases/public/components/user_actions/index.test.tsx:
```diff
@@ -452,18 +356,160 @@ describe(`UserActions`, () => {
+      jest.clearAllMocks();
+    });
+
+    it('Loading spinner when user actions loading and displays fullName/username', () => {
```
nit: let's remove "and displays fullName/username" as we do not check the full name & username.

cnasikas commented on x-pack/plugins/cases/public/components/user_actions/index.test.tsx:
```diff
@@ -452,18 +356,160 @@ describe(`UserActions`, () => {
     });
   });
+
+  describe('pagination', () => {
```
Could you please add a test where we show 20 items (the show more button is not shown) then we add another user action and then check if the show more button shows and the number of total user actions shown is correct?

cnasikas commented on x-pack/plugins/cases/public/components/user_actions/use_user_actions_last_page.test.tsx:
```diff
@@ -0,0 +1,107 @@
+
+const useFindCaseUserActionsMock = useFindCaseUserActions as jest.Mock;
+
+describe('useLastPageUserActions', () => {
```
Can you add a test where you check that `lastPageUserActions` is an empty array when `useFindCaseUserActionsMock` returns `undefined` for data?

cnasikas commented on x-pack/plugins/cases/public/components/user_actions/use_user_actions_pagination.tsx:
```diff
@@ -0,0 +1,72 @@
+        return Math.ceil(userActionsStats.totalOtherActions / perPage);
+      case 'user':
+        return Math.ceil(userActionsStats.totalComments / perPage);
+      case 'all':
+      default:
+        return Math.ceil(userActionsStats.total / perPage);
```
Can you do the following:

```
import { assertNever } from '@elastic/eui';

case 'all':
    return Math.ceil(userActionsStats.total / perPage);
default:
      assertNever(action);
```

This will produce a TS error when we add another `type` in the future and will help us to not forget to update the `switch` statement to include the new type.

cnasikas commented on x-pack/plugins/cases/public/components/user_actions/use_user_actions_pagination.test.tsx:
```diff
@@ -0,0 +1,192 @@
+
+const useInfiniteFindCaseUserActionsMock = useInfiniteFindCaseUserActions as jest.Mock;
+
+describe('useUserActionsPagination', () => {
```
Can you add tests to check the last page for each one of the types?

cnasikas reviewed (commented):
Great job! It is really nice to see this feature 🙂. I left some comments.

js-jankisalvi commented on x-pack/plugins/cases/public/components/case_view/components/case_view_activity.test.tsx:
```diff
@@ -251,104 +259,78 @@ describe.skip('Case View Page activity tab', () => {
       userEvent.click(screen.getByTestId('user-actions-filter-activity-button-comments'));
 
       await waitFor(() => {
-        expect(useFindCaseUserActionsMock).toHaveBeenCalledWith(caseData.id, {
```
removed it to update last page calculation and calls check and then forgot 🤦‍♀️

js-jankisalvi commented on x-pack/plugins/cases/public/containers/use_infinite_find_case_user_actions.test.tsx:
Tried few approaches in [commit](https://github.com/elastic/kibana/pull/152702/commits/4ea9bd28567dcb13d29452d09c4dd637cfc33d58) as per suggestions from [testing-react-query](https://tkdodo.eu/blog/testing-react-query) blog. But it seems I am doing something wrong, as I am not able to `get correct object after fetchNextPage is run`

js-jankisalvi replied:
[Fixed](https://github.com/elastic/kibana/pull/152702/commits/4ce66d3f0fefc97735cd617245e2f7e1469c9ba9)

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/use_last_page.tsx:
```diff
@@ -0,0 +1,41 @@
+import type { CaseUserActionsStats } from '../../containers/types';
+import type { UserActivityParams } from '../user_actions_activity_bar/types';
+
+export const useLastPage = ({
```
open for suggestions for better name 😄

cnasikas reviewed (commented):
Thank you for addressing the comments. I left more comments 🙂. Also. I found out that when pressing the "Show more button" the loading spinner does not show on the button.

js-jankisalvi commented on x-pack/plugins/cases/public/components/case_view/components/case_view_activity.test.tsx:
```diff
@@ -188,32 +209,19 @@ describe.skip('Case View Page activity tab', () => {
 
     const result = appMockRender.render(<CaseViewActivity {...caseProps} />);
     expect(result.getByTestId('case-view-activity')).toBeInTheDocument();
-    expect(result.getByTestId('user-actions')).toBeInTheDocument();
```
hmm.. should be updated.

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/index.tsx:
```diff
@@ -5,251 +5,222 @@
+      data-test-subj="user-actions-loading"
+      isLoading={
+        showBottomList
+          ? isLoadingLastPageUserActions || loadingCommentIds.includes(NEW_COMMENT_ID)
```
I think it had some issues, will check again.

cnasikas replied:
Ok, if it had never mind.

js-jankisalvi replied:
Removed condition from loading as I didn't see any issues anymore.

js-jankisalvi commented on x-pack/plugins/cases/public/components/user_actions/use_last_page.test.tsx:
```diff
@@ -0,0 +1,123 @@
+      })
+    );
+
+    await waitFor(() => {
```
you are right, this one is not waiting on async call.

js-jankisalvi commented:
> I found out that when pressing the "Show more button" the loading spinner does not show on the button.

Because we show user actions and button only after loading is finished 😄

adcoelho commented:
If I have some items at the end of the list and click `Show more` until all are displayed, there will be no margin between the activity items. Maybe you could add some there? 

<details><summary>closed</summary>
<img width="1159" alt="Screenshot 2023-04-03 at 12 39 35" src="https://user-images.githubusercontent.com/1533137/229486823-73507d66-0d90-4a6b-b8fd-aaa7c49ed909.png">
</details>

<details><summary>open (missing margin)</summary>
<img width="1158" alt="Screenshot 2023-04-03 at 12 39 57" src="https://user-images.githubusercontent.com/1533137/229486710-d53e860e-9b33-47a0-ac1a-b8eadb749ebf.png">
</details>

adcoelho commented:
The copy reference link action doesn't work if the item we are trying to link to is inside a page 😬 

At the beginning of the video, I link to an item from the first page, which works fine. It also works if the linked item belongs to the last page.

https://user-images.githubusercontent.com/1533137/229491218-f8dbb011-831b-4f35-92c5-3b6c2c96f6f1.mov

cnasikas commented:
> The copy reference link action doesn't work if the item we are trying to link to is inside a page 😬

Yes, we discussed this and decided to leave it on another PR. @js-jankisalvi Could you please open an issue so we can address it?

js-jankisalvi commented:
> Yes, we discussed this and decided to leave it on another PR. @js-jankisalvi Could you please open an issue so we can address it?

[Created issue](https://github.com/elastic/kibana/issues/154234)

jonathan-buttner reviewed (approved):
Great work @js-jankisalvi !
